crates/swc_ecma_transforms_compat/src/bugfixes/safari_id_destructuring_collision_in_function_expression.rs

```diff
@@ -0,0 +1,129 @@
+            self.fn_expr_name = ident.sym.clone();
+            n.function.params.visit_mut_children_with(self);
+            self.in_body = true;
+            n.function.body.visit_mut_children_with(self);
+            self.in_body = false;
+            if let Some(id_span) = &self.destructured_id_span {
+                let mut rename_map = AHashMap::default();
+                let new_id: JsWord = {
+                    let mut id_value: JsWord = format!("_{}", self.fn_expr_name).into();
+                    let mut count = 0;
+                    while self.other_ident_syms.contains(&id_value) {
+                        count = count + 1;
+                        id_value = format!("_{}{}", self.fn_expr_name, count).into();
+                    }
+                    id_value
+                };
+                let id = (self.fn_expr_name.clone(), id_span.ctxt());
+                rename_map.insert(id, new_id);
+                n.function.visit_mut_children_with(&mut rename(&rename_map));
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use swc_ecma_transforms_testing::test;
+
+    test!(
+        ::swc_ecma_parser::Syntax::default(),
+        |_| safari_id_destructuring_collision_in_function_expression(),
+        basic,
+        "(function a ([a]) { a });
+         (function a({ ...a }) { a });
+         (function a({ a }) { a });",
+        "(function a([_a]) { _a; });
+         (function a({ ..._a }) { _a; });
+         (function a({ a: _a }) { _a; });"
+    );
+
+    test!(
+        ::swc_ecma_parser::Syntax::default(),
+        |_| safari_id_destructuring_collision_in_function_expression(),
+        avoid_collision_1,
+        "(function a([a, _a]) { a + _a })",
+        "(function a([_a1, _a]) {
+          _a1 + _a;
+        });"
+    );
+
+    test!(
+        ::swc_ecma_parser::Syntax::default(),
+        |_| safari_id_destructuring_collision_in_function_expression(),
+        use_duplicated_id,
+        "(function a([a]) { console.log(_a); })",
+        "(function a([_a1]) { console.log(_a); });"
+    );
+
+    test!(
+        ::swc_ecma_parser::Syntax::default(),
+        |_| safari_id_destructuring_collision_in_function_expression(),
+        avoid_collision_2,
+        "(function _a([_a]) { console.log(_a); })",
+        "(function _a([__a]) { console.log(__a); });"
+    );
+
+    test!(
+        ::swc_ecma_parser::Syntax::default(),
+        |_| safari_id_destructuring_collision_in_function_expression(),
+        assign_outside_var,
+        "let _a;
+        (function a([a]) {
+            _a = 3;
+        })",
+        "let _a;
+        (function a([_a1]) {
+            _a = 3;
+        })"
+    );
+
+    test!(
+        ::swc_ecma_parser::Syntax::default(),
+        |_| safari_id_destructuring_collision_in_function_expression(),
+        assignment_expr_in_default_value,
+        "(function a([a = a = 3]) {})",
+        "(function a([_a = _a = 3]) {})"
+    );
+}
```

Comment:
Can you add a test with

input = 
```js
var a = 1;
var _a = 2;

const obj = {
    a: function () {
        a = 2;
        _a = 3;
    },
};
```

and 

pass =

```rust
chain!(resolver(), function_name(), safari_id_destructuring_collision_in_function_expression())
```

?

I think this case will result in a renaming bug. (Sorry for previous wrong change request)